File: nodejs/test/consumer/LiteSimpleConsumer.test.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Offline tests for LiteSimpleConsumer (builder, impl, route filter) and
 * LiteSubscriptionManager. No broker required.
 */

import { describe, it } from 'node:test';
import * as assert from 'node:assert';
import {
  Broker as BrokerPB,
  ClientType,
  LiteSubscriptionAction,
  MessageQueue as MessageQueuePB,
  Permission,
  Settings as SettingsPB,
  Status,
  Code,
  Subscription as SubscriptionPB,
} from '../../proto/apache/rocketmq/v2/definition_pb';
import {
  NotifyUnsubscribeLiteCommand,
  SyncLiteSubscriptionResponse,
} from '../../proto/apache/rocketmq/v2/service_pb';
import { createResource } from '../../src/util';
import { Endpoints, Resource, TopicRouteData } from '../../src/route';
import { LiteSimpleConsumerBuilder } from '../../src/consumer/LiteSimpleConsumer';
import {
  LiteSimpleConsumerImpl,
  firstReadableMasterQueue,
} from '../../src/consumer/LiteSimpleConsumerImpl';
import {
  LiteSubscriptionManager,
  LiteSubscriptionHost,
} from '../../src/consumer/LiteSubscriptionManager';
import { OffsetOption } from '../../src/consumer/OffsetOption';

const ENDPOINTS = '127.0.0.1:8081';

function buildProtoQueue(id: number, brokerId: number, permission: Permission): MessageQueuePB {
  const broker = new BrokerPB()
    .setName(`broker-${brokerId}`)
    .setId(brokerId)
    .setEndpoints(new Endpoints('127.0.0.1:10911').toProtobuf());
  return new MessageQueuePB()
    .setId(id)
    .setTopic(createResource('lite-parent-topic'))
    .setBroker(broker)
    .setPermission(permission);
}

interface SyncRecord {
  endpoints: Endpoints;
  action: LiteSubscriptionAction;
  liteTopics: string[];
}

class FakeHost implements LiteSubscriptionHost {
  clientId = 'fake-client-id';
  running = true;
  quota: number | null = null;
  maxLiteTopicSize: number | null = null;
  readonly endpointsList = [ new Endpoints('127.0.0.1:8081'), new Endpoints('127.0.0.1:8082') ];
  readonly syncRecords: SyncRecord[] = [];
  readonly logger = {
    info: () => undefined,
    warn: () => undefined,
    error: () => undefined,
    debug: () => undefined,
  };

  isRunning(): boolean {
    return this.running;
  }

  getLogger() {
    return this.logger;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  getRpcClientManager(): any {
    return {
      syncLiteSubscription: async (endpoints: Endpoints, request: any) => {
        this.syncRecords.push({
          endpoints,
          action: request.getAction(),
          liteTopics: request.getLiteTopicSetList(),
        });
        return new SyncLiteSubscriptionResponse().setStatus(new Status().setCode(Code.OK));
      },
    };
  }

  getRequestTimeout(): number {
    return 3000;
  }

  getSyncEndpoints(): Endpoints[] {
    return this.endpointsList;
  }

  resetRecords() {
    this.syncRecords.length = 0;
  }
}

describe('test/consumer/LiteSimpleConsumer.test.ts', () => {

  describe('LiteSimpleConsumerBuilder', () => {
    it('should chain bindTopic and validate blank topic', () => {
      const builder = new LiteSimpleConsumerBuilder();
      assert.strictEqual(builder.bindTopic('test-bind-topic'), builder);
      assert.throws(() => builder.bindTopic(''), /bindTopic should not be blank/);
      assert.throws(() => builder.bindTopic('   '), /bindTopic should not be blank/);
    });

    it('should validate consumer group', () => {
      const builder = new LiteSimpleConsumerBuilder();
      assert.throws(() => builder.setConsumerGroup(null as any), /consumerGroup should not be null/);
      assert.throws(() => builder.setConsumerGroup('invalid group!'), /does not match the pattern/);
      assert.strictEqual(builder.setConsumerGroup('valid-group_1'), builder);
    });

    it('should validate await duration', () => {
      const builder = new LiteSimpleConsumerBuilder();
      assert.throws(() => builder.setAwaitDuration(0), /awaitDuration should be positive/);
      assert.throws(() => builder.setAwaitDuration(-1), /awaitDuration should be positive/);
      assert.strictEqual(builder.setAwaitDuration(5000), builder);
    });

    it('should fail to build without required options', async () => {
      await assert.rejects(async () => {
        await new LiteSimpleConsumerBuilder().build();
      }, /clientConfiguration has not been set yet/);

      await assert.rejects(async () => {
        await new LiteSimpleConsumerBuilder().setClientConfiguration({ endpoints: ENDPOINTS, namespace: '' }).build();
      }, /consumerGroup has not been set yet/);

      await assert.rejects(async () => {
        await new LiteSimpleConsumerBuilder()
          .setClientConfiguration({ endpoints: ENDPOINTS, namespace: '' })
          .setConsumerGroup('valid-group')
          .build();
      }, /bindTopic has not been set yet/);
    });
  });

  describe('LiteSimpleConsumerImpl constructor', () => {
    const baseOptions = {
      endpoints: ENDPOINTS,
      namespace: '',
      consumerGroup: 'lite-unittest-group',
      bindTopic: 'lite-parent-topic',
    };

    it('should reject blank bindTopic', () => {
      assert.throws(() => {
        new LiteSimpleConsumerImpl({ ...baseOptions, bindTopic: ' ' });
      }, /bindTopic should not be blank/);
    });

    it('should use LITE_SIMPLE_CONSUMER client type', () => {
      const impl = new LiteSimpleConsumerImpl(baseOptions);
      assert.strictEqual((impl as any).getClientType(), ClientType.LITE_SIMPLE_CONSUMER);
    });

    it('should subscribe the bind topic with SUB_ALL by default', () => {
      const impl = new LiteSimpleConsumerImpl(baseOptions);
      const settings = (impl as any).getSettings();
      const protobuf = settings.toProtobuf();
      const subscription = protobuf.getSubscription()!;
      assert.strictEqual(subscription.getGroup()!.getName(), 'lite-unittest-group');
      const entries = subscription.getSubscriptionsList();
      assert.strictEqual(entries.length, 1);
      assert.strictEqual(entries[0].getTopic()!.getName(), 'lite-parent-topic');
      assert.strictEqual(entries[0].getExpression()!.getExpression(), '*');
      assert.strictEqual(protobuf.getClientType(), ClientType.LITE_SIMPLE_CONSUMER);
    });

    it('should expose group and empty lite topic set before startup', () => {
      const impl = new LiteSimpleConsumerImpl(baseOptions);
      assert.strictEqual(impl.getConsumerGroup(), 'lite-unittest-group');
      assert.strictEqual(impl.getLiteTopicSet().size, 0);
    });
  });

  describe('firstReadableMasterQueue', () => {
    it('should keep only the first readable master queue', () => {
      const routeData = new TopicRouteData([
        buildProtoQueue(0, 1, Permission.READ_WRITE), // slave readable
        buildProtoQueue(1, 0, Permission.WRITE), // master not readable
        buildProtoQueue(2, 0, Permission.READ_WRITE), // first readable master
        buildProtoQueue(3, 0, Permission.READ), // second readable master
      ]);

      const filtered = firstReadableMasterQueue(routeData);
      assert.strictEqual(filtered.messageQueues.length, 1);
      assert.strictEqual(filtered.messageQueues[0].broker.id, 0);
      assert.strictEqual(filtered.messageQueues[0].queueId, 2);
    });

    it('should return empty route when no readable master exists', () => {
      const routeData = new TopicRouteData([
        buildProtoQueue(0, 0, Permission.WRITE),
        buildProtoQueue(1, 1, Permission.READ),
      ]);

      const filtered = firstReadableMasterQueue(routeData);
      assert.strictEqual(filtered.messageQueues.length, 0);
    });
  });

  describe('LiteSubscriptionManager', () => {
    const bindTopic = 'lite-parent-topic';

    function createManager(host = new FakeHost()) {
      const manager = new LiteSubscriptionManager(
        host,
        new Resource('', bindTopic),
        new Resource('', 'lite-unittest-group'),
      );
      return { host, manager };
    }

    it('should sync PARTIAL_ADD to every sync endpoint on subscribeLite', async () => {
      const { host, manager } = createManager();
      await manager.subscribeLite('lite-topic-1', OffsetOption.MIN_OFFSET);

      assert.strictEqual(host.syncRecords.length, 2);
      for (const record of host.syncRecords) {
        assert.strictEqual(record.action, LiteSubscriptionAction.PARTIAL_ADD);
        assert.deepStrictEqual(record.liteTopics, [ 'lite-topic-1' ]);
      }
      assert.ok(host.syncRecords.some(r => r.endpoints.facade === '127.0.0.1:8081'));
      assert.ok(host.syncRecords.some(r => r.endpoints.facade === '127.0.0.1:8082'));
      assert.ok(manager.getLiteTopicSet().has('lite-topic-1'));
      assert.strictEqual(manager.getBindTopicName(), bindTopic);
      assert.strictEqual(manager.getConsumerGroupName(), 'lite-unittest-group');
    });

    it('should reject subscribeLite when not running', async () => {
      const { host, manager } = createManager();
      host.running = false;
      await assert.rejects(async () => {
        await manager.subscribeLite('lite-topic-1');
      }, /Consumer is not running/);
      assert.strictEqual(host.syncRecords.length, 0);
    });

    it('should skip duplicate subscribeLite without extra rpc', async () => {
      const { host, manager } = createManager();
      await manager.subscribeLite('lite-topic-1');
      host.resetRecords();
      await manager.subscribeLite('lite-topic-1');
      assert.strictEqual(host.syncRecords.length, 0);
      assert.strictEqual(manager.getLiteTopicSet().size, 1);
    });

    it('should validate lite topic name and length', async () => {
      const { manager } = createManager();
      await assert.rejects(async () => {
        await manager.subscribeLite('   ');
      }, /liteTopic is blank/);
      await assert.rejects(async () => {
        await manager.subscribeLite('x'.repeat(65));
      }, /liteTopic length exceeded max length 64/);
    });

    it('should throw when quota exceeded', async () => {
      const { manager } = createManager();
      const settings = new SettingsPB()
        .setSubscription(new SubscriptionPB().setLiteSubscriptionQuota(1));
      manager.sync(settings);
      await manager.subscribeLite('lite-topic-1');
      await assert.rejects(async () => {
        await manager.subscribeLite('lite-topic-2');
      }, /Lite subscription quota exceeded 1/);
      assert.strictEqual(manager.getLiteTopicSet().size, 1);
    });

    it('should adopt maxLiteTopicSize from settings', async () => {
      const { manager } = createManager();
      const settings = new SettingsPB()
        .setSubscription(new SubscriptionPB().setMaxLiteTopicSize(8));
      manager.sync(settings);
      await assert.rejects(async () => {
        await manager.subscribeLite('x'.repeat(9));
      }, /liteTopic length exceeded max length 8/);
    });

    it('should sync PARTIAL_REMOVE on unsubscribeLite', async () => {
      const { host, manager } = createManager();
      await manager.subscribeLite('lite-topic-1');
      host.resetRecords();

      await manager.unsubscribeLite('lite-topic-1');
      assert.strictEqual(host.syncRecords.length, 2);
      assert.ok(host.syncRecords.every(r => r.action === LiteSubscriptionAction.PARTIAL_REMOVE));
      assert.strictEqual(manager.getLiteTopicSet().size, 0);
    });

    it('should skip unsubscribeLite for unknown topic', async () => {
      const { host, manager } = createManager();
      await manager.unsubscribeLite('never-subscribed');
      assert.strictEqual(host.syncRecords.length, 0);
    });

    it('should fail subscription when any endpoint rejects the sync', async () => {
      const host = new FakeHost();
      const manager = new LiteSubscriptionManager(
        host,
        new Resource('', bindTopic),
        new Resource('', 'lite-unittest-group'),
      );
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (host as any).getRpcClientManager = () => ({
        syncLiteSubscription: async (endpoints: Endpoints) => {
          if (endpoints.facade === '127.0.0.1:8081') {
            return new SyncLiteSubscriptionResponse().setStatus(
              new Status().setCode(Code.INTERNAL_SERVER_ERROR),
            );
          }
          return new SyncLiteSubscriptionResponse().setStatus(new Status().setCode(Code.OK));
        },
      });

      await assert.rejects(async () => {
        await manager.subscribeLite('lite-topic-1');
      }, /Failed to sync lite subscription/);
      // Failed sync must not pollute the local set
      assert.strictEqual(manager.getLiteTopicSet().size, 0);
    });

    it('should remove lite topic on notify unsubscribe command', async () => {
      const { manager } = createManager();
      await manager.subscribeLite('lite-topic-1');
      const command = new NotifyUnsubscribeLiteCommand().setLiteTopic('lite-topic-1');
      manager.onNotifyUnsubscribeLiteCommand(command);
      assert.strictEqual(manager.getLiteTopicSet().size, 0);
    });

    it('should ignore blank lite topic in notify unsubscribe command', async () => {
      const { manager } = createManager();
      await manager.subscribeLite('lite-topic-1');
      manager.onNotifyUnsubscribeLiteCommand(new NotifyUnsubscribeLiteCommand().setLiteTopic(''));
      assert.ok(manager.getLiteTopicSet().has('lite-topic-1'));
    });

    it('should shutdown cleanly and clear the topic set', async () => {
      const { manager } = createManager();
      await manager.subscribeLite('lite-topic-1');
      manager.startUp();
      manager.shutdown();
      assert.strictEqual(manager.getLiteTopicSet().size, 0);
      manager.shutdown(); // idempotent
    });
  });
});

File: csharp/tests/LiteSimpleConsumerTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Moq;
using Org.Apache.Rocketmq;
using Org.Apache.Rocketmq.Error;
using Proto = Apache.Rocketmq.V2;

namespace tests
{
    [TestClass]
    public class LiteSimpleConsumerTest
    {
        private const string BindTopic = "lite-parent-topic";
        private const string ConsumerGroup = "lite-simple-consumer-group";

        [TestMethod]
        public void TestBuildWithoutClientConfig()
        {
            var builder = new LiteSimpleConsumer.Builder().SetConsumerGroup(ConsumerGroup).SetBindTopic(BindTopic);
            Assert.ThrowsExactly<ArgumentException>(() => builder.SetClientConfig(null));
        }

        [TestMethod]
        public void TestBuildWithInvalidConsumerGroup()
        {
            Assert.ThrowsExactly<ArgumentException>(() =>
                new LiteSimpleConsumer.Builder().SetConsumerGroup("invalid group!"));
            Assert.ThrowsExactly<ArgumentException>(() =>
                new LiteSimpleConsumer.Builder().SetConsumerGroup(null));
        }

        [TestMethod]
        public void TestBuildWithBlankBindTopic()
        {
            Assert.ThrowsExactly<ArgumentException>(() => new LiteSimpleConsumer.Builder().SetBindTopic(""));
        }

        [TestMethod]
        public void TestSetNonPositiveAwaitDuration()
        {
            Assert.ThrowsExactly<ArgumentException>(() =>
                new LiteSimpleConsumer.Builder().SetAwaitDuration(TimeSpan.Zero));
        }

        [TestMethod]
        public void TestBuildWithoutRequiredOptions()
        {
            var builder = new LiteSimpleConsumer.Builder();
            Assert.ThrowsExactly<ArgumentException>(() => builder.Build().GetAwaiter().GetResult());
        }

        [TestMethod]
        public void TestConstructorRejectsBlankBindTopic()
        {
            var clientConfig = new ClientConfig.Builder().SetEndpoints("127.0.0.1:8081").Build();
            Assert.ThrowsExactly<ArgumentException>(() =>
                new LiteSimpleConsumer(clientConfig, ConsumerGroup, TimeSpan.FromSeconds(5), "  "));
        }

        [TestMethod]
        public void TestSettingsUseLiteSimpleConsumerClientType()
        {
            var consumer = CreateConsumer(out _);
            var settings = consumer.GetSettings().ToProtobuf();
            Assert.AreEqual(Proto.ClientType.LiteSimpleConsumer, settings.ClientType);
            Assert.IsTrue(consumer.IsLiteConsumer());
        }

        [TestMethod]
        public void TestHeartbeatUsesLiteSimpleConsumerClientType()
        {
            var consumer = CreateConsumer(out _);
            Assert.AreEqual(Proto.ClientType.LiteSimpleConsumer, consumer.WrapHeartbeatRequest().ClientType);
        }

        [TestMethod]
        public async Task TestSubscribeLiteBeforeStartup()
        {
            var clientConfig = new ClientConfig.Builder().SetEndpoints("127.0.0.1:8081").Build();
            var consumer = new LiteSimpleConsumer(clientConfig, ConsumerGroup, TimeSpan.FromSeconds(5), BindTopic);
            await Assert.ThrowsExactlyAsync<InvalidOperationException>(async () =>
                await consumer.SubscribeLite("lite-topic-1"));
        }

        [TestMethod]
        public async Task TestSubscribeLiteSyncsPartialAdd()
        {
            var consumer = CreateConsumer(out var mockClientManager);
            var invocations = SetupSyncLiteSubscription(mockClientManager);

            await consumer.SubscribeLite("lite-topic-1", OffsetOption.MinOffset);

            Assert.AreEqual(1, invocations.Count);
            var request = invocations[0];
            Assert.AreEqual(Proto.LiteSubscriptionAction.PartialAdd, request.Action);
            CollectionAssert.Contains(request.LiteTopicSet.ToList(), "lite-topic-1");
            Assert.AreEqual(BindTopic, request.Topic.Name);
            Assert.AreEqual(ConsumerGroup, request.Group.Name);
            Assert.AreEqual(Proto.OffsetOption.Types.Policy.Min, request.OffsetOption.Policy);
            Assert.IsTrue(consumer.GetLiteTopicSet().Contains("lite-topic-1"));
        }

        [TestMethod]
        public async Task TestDuplicateSubscribeLiteSkipsRpc()
        {
            var consumer = CreateConsumer(out var mockClientManager);
            var invocations = SetupSyncLiteSubscription(mockClientManager);

            await consumer.SubscribeLite("lite-topic-1");
            await consumer.SubscribeLite("lite-topic-1");

            Assert.AreEqual(1, invocations.Count);
        }

        [TestMethod]
        public async Task TestUnsubscribeLiteSyncsPartialRemove()
        {
            var consumer = CreateConsumer(out var mockClientManager);
            var invocations = SetupSyncLiteSubscription(mockClientManager);

            await consumer.SubscribeLite("lite-topic-1");
            await consumer.UnsubscribeLite("lite-topic-1");

            Assert.AreEqual(2, invocations.Count);
            Assert.AreEqual(Proto.LiteSubscriptionAction.PartialRemove, invocations[1].Action);
            Assert.AreEqual(0, consumer.GetLiteTopicSet().Count);
        }

        [TestMethod]
        public async Task TestUnsubscribeUnknownLiteTopicIsNoop()
        {
            var consumer = CreateConsumer(out var mockClientManager);
            var invocations = SetupSyncLiteSubscription(mockClientManager);

            await consumer.UnsubscribeLite("unknown-lite-topic");

            Assert.AreEqual(0, invocations.Count);
        }

        [TestMethod]
        public async Task TestSubscribeLiteWithBlankTopic()
        {
            var consumer = CreateConsumer(out var mockClientManager);
            SetupSyncLiteSubscription(mockClientManager);

            await Assert.ThrowsExactlyAsync<ArgumentException>(async () => await consumer.SubscribeLite("  "));
        }

        [TestMethod]
        public async Task TestOnNotifyUnsubscribeLiteCommand()
        {
            var consumer = CreateConsumer(out var mockClientManager);
            SetupSyncLiteSubscription(mockClientManager);
            await consumer.SubscribeLite("lite-topic-1");

            consumer.OnNotifyUnsubscribeLiteCommand(null, new Proto.NotifyUnsubscribeLiteCommand { LiteTopic = "lite-topic-1" });
            Assert.AreEqual(0, consumer.GetLiteTopicSet().Count);

            await consumer.SubscribeLite("lite-topic-2");
            consumer.OnNotifyUnsubscribeLiteCommand(null, new Proto.NotifyUnsubscribeLiteCommand { LiteTopic = "" });
            Assert.IsTrue(consumer.GetLiteTopicSet().Contains("lite-topic-2"));
        }

        [TestMethod]
        public void TestAckRequestCarriesLiteTopic()
        {
            var consumer = CreateConsumer(out _);
            var messageView = CreateMessageView("lite-topic-1");

            var request = consumer.WrapAckMessageRequest(messageView);

            Assert.AreEqual(1, request.Entries.Count);
            Assert.AreEqual("lite-topic-1", request.Entries[0].LiteTopic);
            Assert.AreEqual(BindTopic, request.Topic.Name);
        }

        [TestMethod]
        public void TestAckRequestWithoutLiteTopic()
        {
            var consumer = CreateConsumer(out _);
            var messageView = CreateMessageView(null);

            var request = consumer.WrapAckMessageRequest(messageView);

            Assert.IsFalse(request.Entries[0].HasLiteTopic);
        }

        [TestMethod]
        public void TestChangeInvisibleDurationCarriesLiteTopic()
        {
            var consumer = CreateConsumer(out _);
            var messageView = CreateMessageView("lite-topic-1");

            var request = consumer.WrapChangeInvisibleDuration(messageView, TimeSpan.FromSeconds(10));

            Assert.AreEqual("lite-topic-1", request.LiteTopic);
        }

        [TestMethod]
        public void TestUpdateSubscriptionLoadBalancerKeepsOnlyFirstReadableMasterQueue()
        {
            var consumer = CreateConsumer(out _);
            var routeData = new TopicRouteData(new List<Proto.MessageQueue>
            {
                CreateMessageQueue(0, 0, Proto.Permission.ReadWrite),
                CreateMessageQueue(1, 0, Proto.Permission.ReadWrite),
            });

            var loadBalancer = consumer.UpdateSubscriptionLoadBalancer(BindTopic, routeData);

            Assert.IsNotNull(loadBalancer);
            var mq = loadBalancer.TakeMessageQueue();
            Assert.AreEqual(0, mq.QueueId);
        }

        [TestMethod]
        public void TestUpdateSubscriptionLoadBalancerWithoutReadableMasterQueue()
        {
            var consumer = CreateConsumer(out _);
            var routeData = new TopicRouteData(new List<Proto.MessageQueue>
            {
                CreateMessageQueue(0, 1, Proto.Permission.ReadWrite),
            });

            var loadBalancer = consumer.UpdateSubscriptionLoadBalancer(BindTopic, routeData);

            Assert.ThrowsExactly<NotFoundException>(() => loadBalancer.TakeMessageQueue());
        }

        private static LiteSimpleConsumer CreateConsumer(out Mock<IClientManager> mockClientManager)
        {
            var clientConfig = new ClientConfig.Builder().SetEndpoints("127.0.0.1:8081").Build();
            var consumer = new LiteSimpleConsumer(clientConfig, ConsumerGroup, TimeSpan.FromSeconds(5), BindTopic);
            mockClientManager = new Mock<IClientManager>();
            consumer.SetClientManager(mockClientManager.Object);
            consumer.State = State.Running;
            // The lite subscription quota is pushed by the server through the settings command.
            consumer.OnSettingsCommand(null, new Proto.Settings
            {
                Subscription = new Proto.Subscription
                {
                    LiteSubscriptionQuota = 64,
                    MaxLiteTopicSize = 64
                }
            });
            return consumer;
        }

        private static List<Proto.SyncLiteSubscriptionRequest> SetupSyncLiteSubscription(Mock<IClientManager> mockClientManager)
        {
            var invocations = new List<Proto.SyncLiteSubscriptionRequest>();
            mockClientManager.Setup(cm => cm.SyncLiteSubscription(It.IsAny<Endpoints>(),
                    It.IsAny<Proto.SyncLiteSubscriptionRequest>(), It.IsAny<TimeSpan>()))
                .Returns((Endpoints _, Proto.SyncLiteSubscriptionRequest request, TimeSpan _) =>
                {
                    invocations.Add(request.Clone());
                    var response = new Proto.SyncLiteSubscriptionResponse
                    {
                        Status = new Proto.Status { Code = Proto.Code.Ok }
                    };
                    return Task.FromResult(
                        new RpcInvocation<Proto.SyncLiteSubscriptionRequest, Proto.SyncLiteSubscriptionResponse>(
                            request, response, null));
                });
            return invocations;
        }

        private static MessageView CreateMessageView(string liteTopic)
        {
            var systemProperties = new Proto.SystemProperties
            {
                MessageType = Proto.MessageType.Normal,
                MessageId = MessageIdGenerator.GetInstance().Next(),
                BodyDigest = new Proto.Digest { Type = Proto.DigestType.Crc32, Checksum = "9EF61F95" },
                BodyEncoding = Proto.Encoding.Identity,
                BornHost = "127.0.0.1",
                BornTimestamp = new Timestamp(),
                ReceiptHandle = "fake-receipt-handle",
            };
            if (null != liteTopic)
            {
                systemProperties.LiteTopic = liteTopic;
            }

            var message = new Proto.Message
            {
                SystemProperties = systemProperties,
                Topic = new Proto.Resource { Name = BindTopic },
                Body = ByteString.CopyFrom("foobar", Encoding.UTF8)
            };
            return MessageView.FromProtobuf(message);
        }

        private static Proto.MessageQueue CreateMessageQueue(int queueId, int brokerId, Proto.Permission permission)
        {
            return new Proto.MessageQueue
            {
                Id = queueId,
                Permission = permission,
                Broker = new Proto.Broker
                {
                    Id = brokerId,
                    Name = "broker0",
                    Endpoints = new Proto.Endpoints
                    {
                        Scheme = Proto.AddressScheme.Ipv4,
                        Addresses = { new Proto.Address { Host = "127.0.0.1", Port = 8081 } }
                    }
                },
                Topic = new Proto.Resource { Name = BindTopic },
                AcceptMessageTypes = { Proto.MessageType.Lite }
            };
        }
    }
}
